lib/shopify_app/session/actual_session_storage.rb

```diff
@@ -0,0 +1,44 @@
+# frozen_string_literal: true
+module ShopifyApp
+  module ActualSessionStorage
+    extend ActiveSupport::Concern
+    include ::ShopifyApp::SessionStorage
+
+    class_methods do
+      def store(auth_session, session_id, user, expires_at)
```

Comment:
I wonder if you can simplify the interface here.

`expires_at` is a property of the token, maybe it should live on the `ShopifyAPI::Session` object?

And `session_id` is a property of the user session, and we already have an object that represents that: `user`.

If you can simplify the interface in this way, then the interface will be the same as the interface for the user session storage strategy. Then this can just be another user session storage! That will simplify things a lot, and make migrating easier (since they can just switch to the new session storage when they're ready).

Reply:
I understand what you mean about grouping `expires_at` within the `ShopifyAPI::Session` object. 

For the `session_id` to be within `user` do you mean that I should make an object called `user` that wraps the `associated_user` and `jwt_shopify_session_id` within the `CallbackController`? 

And then send this information like `ShopifyApp::SessionRepository.store_actual_session(session_store, user)`

Reply:
I was kind of sloppily making two points at once:
1. We should try to implement all the aspects of the concept in the same object. This makes the system easier to maintain, because when all of the aspects of a concept are represented in a single object, we can often implement logic that only the object needs to understand the details about, and the rest of the system only needs to know the interface that the object provides. "Low coupling, high cohesion" is one way developers talk about this principle.
2. If we don't have to add a new concept, then we shouldn't! The easiest code to maintain is the code we never write. I was suggesting that we can probably make the interface for the existing user session storage the same as the interface for the "actual session" that you're proposing. Then we don't need the new "actual session" interface!

I'm less sure that 2. is true. For example, we usually, but not always, have a session ID in the "normal" OAuth, so there's a way to populate that value in the interface. It's this value: https://github.com/Shopify/shopify_app/blob/master/app/controllers/shopify_app/callback_controller.rb#L133

What do we do when we don't have that value? If our class permits the value not to exist, then we have to handle that case. You might decide that it's not acceptable to be missing the session ID in the "actual session" case, in which case we can't support the same interface. I'd encourage you to think about a fallback when we don't have a session value though (and maybe logging the event), because I think it will simplify the maintainability quite a bit and be easier for partners to understand as they migrate.

Re: how to handle `session_id` and the `user` object, I wouldn't wrap it, I'd define a real class that encapsulates the information about the user session. So maybe something like `ShopifyApp::UserSession`. Then you can construct it using data from these different sources.